### PR TITLE
docs(api-service): Docs for api service

### PR DESCRIPTION
**Pull request by harshthakur9030**
Signed-off-by: Harsh Thakur <harshthakur9030@gmail.com>

## Pull Request template

**Why is this PR required? What issue does it fix?**:
Add docs for api-service

**What this PR does?**:
Add docs for api-service

**Does this PR require any upgrade changes?**:
No

**If the changes in this PR are manually verified, list down the scenarios covered:**:

**Any additional information for your reviewer?** : 
 #433 

**Checklist:**
- [ ] Fixes #<issue number>
- [x] PR Title follows the convention of  `<type>(<scope>): <subject>`
- [ ] Has the change log section been updated? 
- [ ] Commit has unit tests
- [ ] Commit has integration tests
- [ ] (Optional) Are upgrade changes included in this PR? If not, mention the issue/PR to track: 
- [ ] (Optional) If documentation changes are required, which issue on https://github.com/openebs/openebs-docs is used to track them:

### CODE REVIEW

File: docs/api-service-user.md [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+- Block Device Details:  Given the block device path, it can give [S.M.A.R.T](https://en.wikipedia.org/wiki/S.M.A.R.T.) stats of that block device. 
+
+- iSCSI status : Since most of the data engines supported by OpenEBS require iSCSI, it's status can be checked from here. 
+
+- Hugepages : Mayastor requires hugepages to be set on the node. Hence, you can use the API to set hugepages and check the number of hugepages available on the node. 
+**Note : Running the SET method does **NOT** guarantee that the hugepages will be set, it's best to verify it by using the GET method.**
+
+- Rescan : Instead of restarting the NDM pod when something seems wrong, Rescan can be run to sync NDM's local state with etcd. NDM primarily relies on [udev](https://opensource.com/article/18/11/udev) events and rarely, the events can be missed by NDM.
+ For instance, a disk could be attached to the node but you don't see it in output of `kubectl get bd` . In such a case, running a rescan would help. Rescan doesn't help when there's a change in capacity of the block device and when a filesystem is mounted, this is because udev doesn't generate events for those.
```
Comment (akhilerm) [inline docs/api-service-user.md:22]:
```suggestion
 For instance, a disk could be attached to the node but you don't see it in output of `kubectl get bd` . In such a case, running a rescan would help. Rescan also helps when there's a change in capacity of the block device or when a filesystem is mounted. All the devices will be scanned and latest information will be updated to etcd.
```

File: docs/api-service-dev.md [hunk 2]
```diff
@@ -0,0 +1,13 @@
+## API service developer guide
+
+# Steps :
+
+- Use a proto lint extension, if you're using VS code, use [this](https://marketplace.visualstudio.com/items?itemName=zxh404.vscode-proto3).
```
Comment (akhilerm) [inline docs/api-service-dev.md:5]:
```suggestion
- Developers can use a proto lint extension while making changes to the proto file.
eg: if you're using VS code, use [vscode-proto3](https://marketplace.visualstudio.com/items?itemName=zxh404.vscode-proto3).
```

File: docs/api-service-dev.md [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+
+- After any change in [proto file](../ndm.proto), run:
```
Comment (akhilerm) [inline docs/api-service-dev.md:7]:
```suggestion
- After any change in the [proto file](../ndm.proto), the corresponding `golang` code should be generated. To generate run: 
```

File: docs/api-service-dev.md [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+
+# Steps :
+
+- Use a proto lint extension, if you're using VS code, use [this](https://marketplace.visualstudio.com/items?itemName=zxh404.vscode-proto3).
+
+- After any change in [proto file](../ndm.proto), run: 
+
+```make protos```
+
+- Generated code can be found under [spec/ndm](../spec/ndm/ndm.pb.go)
```
Comment (akhilerm) [inline docs/api-service-dev.md:11]:
```suggestion
- The generated code can be found under [spec/ndm](../spec/ndm/ndm.pb.go)
```

File: docs/api-service-dev.md [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+
+- Code related to api service should be done under api-service folder. Based on the functionality, it can go under the node or cluster folder.
```
Comment (akhilerm) [inline docs/api-service-dev.md:13]:
```suggestion
- The code related new functionality for the API service should be added under the `api-service` directory. Based on the type of functionality added, it should be packaged as follows:
    - cluster: all services that are common to the cluster, like version of NDM.
    - node: services that are node dependent, like list of blockdevices, checking iSCSI status etc.
```

File: docs/api-service-user.md [hunk 6]
```diff
@@ -0,0 +1,25 @@
+# API service user guide
+
+## What does this API do?
+It exposes functionality of NDM which can help in  gaining information about block devices and troubleshooting over gRPC.
```
Comment (akhilerm) [inline docs/api-service-user.md:4]:
```suggestion
The API service exposes some functionality of NDM to the user which can help in gaining additional information about block devices and troubleshooting. It is exposed as a gRPC service.
```

File: docs/api-service-user.md [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+
+## Currently exposed services:
```
Comment (akhilerm) [inline docs/api-service-user.md:6]:
```suggestion
## Currently available services:
```

File: docs/api-service-user.md [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+- Node name : Gives the name of the node it is running on
+
+- List Block Devices : Similar to `lsblk`, it lists all the block devices it finds and relationships among them.
+
+- Block Device Details:  Given the block device path, it can give [S.M.A.R.T](https://en.wikipedia.org/wiki/S.M.A.R.T.) stats of that block device. 
+
+- iSCSI status : Since most of the data engines supported by OpenEBS require iSCSI, it's status can be checked from here. 
+
+- Hugepages : Mayastor requires hugepages to be set on the node. Hence, you can use the API to set hugepages and check the number of hugepages available on the node. 
+**Note : Running the SET method does **NOT** guarantee that the hugepages will be set, it's best to verify it by using the GET method.**
```
Comment (akhilerm) [inline docs/api-service-user.md:19]:
```suggestion
**Note : Running the SET method does NOT guarantee that the hugepages will be set, it's recommended to verify by using the GET method.**
```

File: docs/api-service-user.md [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+
+- Rescan : Instead of restarting the NDM pod when something seems wrong, Rescan can be run to sync NDM's local state with etcd. NDM primarily relies on [udev](https://opensource.com/article/18/11/udev) events and rarely, the events can be missed by NDM.
+ For instance, a disk could be attached to the node but you don't see it in output of `kubectl get bd` . In such a case, running a rescan would help. Rescan doesn't help when there's a change in capacity of the block device and when a filesystem is mounted, this is because udev doesn't generate events for those.  
+
+## How to use it?
+CLI is in the works, meanwhile a client like [grpcurl](https://github.com/fullstorydev/grpcurl) can be used.
```
Comment (akhilerm) [inline docs/api-service-user.md:25]:
```suggestion
CLI for accessing the service is not completely implemented. A client like [grpcurl](https://github.com/fullstorydev/grpcurl) can be used currently to access the gRPC service.
```

File: docs/api-service-user.md (hunk 1 above)
Comment (harshthakur9030) [reply docs/api-service-user.md:22]: I thought it doesn't help because udev doesn't generate events for those. Did I get that wrong?

Comment (akhilerm) [reply docs/api-service-user.md:22]: udev doesn't generate events for those. That is correct. But if we rescan, along with udev we also fetch the current capacity and mountpoint. So rescan actually helps to get lhe changed size.

**Comment by akhilerm** [comment]
Please check the DCO, and put a reference to these docs from the developer docs or contributing file.

File: CONTRIBUTING.md [hunk 10]
```diff
@@ -130,5 +130,8 @@ This project is implemented using Go and uses the standard golang tools for deve
 
 For setting up a Development environment on your local host, see the detailed instructions [here](./BUILD.md).
 
-The NDM design document is available [here](./docs/design.md).
-
+## Design documents
```
Comment (akhilerm) [inline CONTRIBUTING.md:133]:
```suggestion
## Design documents
The following are the design documents and flow diagrams of major components in NDM:
```

File: CONTRIBUTING.md [hunk 11]
```diff
@@ -130,5 +130,8 @@ This project is implemented using Go and uses the standard golang tools for deve
...
+- The NDM design document is available [here](./docs/design.md).
+- Cleanup design doc is available [here](./docs/cleanup-design.md).
+- Exporter design doc is available [here](./docs/exporter.md).
+- API service design doc is available [here](./docs/api-service-dev.md).
```
Comment (akhilerm) [inline CONTRIBUTING.md:137]:
```suggestion
- [NDM design doc](./docs/design.md).
- [Disk cleanup design doc](./docs/cleanup-design.md).
- [NDM Exporter design doc](./docs/exporter.md).
- [API service design doc](./docs/api-service-dev.md).
```

**Pull request merged**